pkg/mutation/path/parser/node.go

```diff
@@ -66,6 +67,118 @@ func (r Path) String() string {
 	return result.String()
 }
 
+func (r Path) GetValueAtLocation(object interface{}) ([]string, error) {
+	var recurse func(interface{}, int) ([]string, error)
+	recurse = func(current interface{}, depth int) ([]string, error) {
+		pathEntry := r.Nodes[depth]
+		switch castPathEntry := pathEntry.(type) {
+		case *Object:
+			obj, ok := current.(map[string]interface{})
+			if !ok {
+				return nil, fmt.Errorf("mismatch between path entry (type: object) and received object (type: %T). Path: %+v", current, castPathEntry)
+			}
+
+			next, ok := obj[castPathEntry.Reference]
+			if !ok {
+				return nil, fmt.Errorf("path entry does not exist in current object. Path: %+v", castPathEntry)
+			}
+
+			// we have hit the end of our path, this is the base case
+			// value at location can be a plain value, an array, or a composite value
+			// for array and composite value, we should json marshal it
+			// otherwise, return the string representation of the inteface
+			if depth == len(r.Nodes)-1 {
+				switch cast := next.(type) {
+				case map[string]interface{}, []interface{}:
+					// convert composite value or array into json string representation
+					bytes, err := json.Marshal(cast)
+					if err != nil {
+						return nil, fmt.Errorf("failed to marshal value: %w", err)
+					}
+					return []string{string(bytes)}, nil
+				default:
+					// get string representation of any type
+					return []string{fmt.Sprintf("%v", cast)}, nil
+				}
+			}
+
+			return recurse(next, depth+1)
+		case *List:
+			list, ok := current.([]interface{})
+			if !ok {
+				return nil, fmt.Errorf("mismatch between path entry (type: List) and received object (type: %T). Path: %+v", current, castPathEntry)
+			}
+
+			glob := castPathEntry.Glob
+			key, expectedValue := castPathEntry.KeyField, castPathEntry.KeyValue
+
+			// we have hit the end of our path, this is the base case
+			if depth == len(r.Nodes)-1 {
+				if glob {
+					return nil, fmt.Errorf("last path entry can not be globbed")
+				}
+
+				var lastItem interface{}
+				for _, v := range list {
+					switch cast := v.(type) {
+					case map[string]interface{}:
+						v, ok := cast[key]
+						if !ok || v != expectedValue {
+							continue
+						}
+						lastItem = cast
+					default:
+						// get string representation of any type
+						lastItem = fmt.Sprintf("%v", cast)
+					}
+				}
+				if lastItem == nil {
+					return nil, fmt.Errorf("no matching value found")
+				}
+
+				bytes, err := json.Marshal(lastItem)
+				if err != nil {
+					return nil, fmt.Errorf("failed to marshal value: %w", err)
+				}
+
+				return []string{string(bytes)}, nil
+			}
+
+			if glob {
+				values := make([]string, len(list))
+				for i, item := range list {
+					potentialValues, err := recurse(item, depth+1)
+					if err != nil {
+						return nil, fmt.Errorf("failed to recurse: %w", err)
+					}
+					// we can assume that potentialValues contains only one value
```

Comment:
Can we assume potentialValues contains only one value?

re-raising the example of `spec.containers[name: *].image` where:

```yaml
spec:
  containers:
  - name: "foo"
    image: "my-docker-image"
  - name: "bar"
    image: "my-sidecar-image"
```

what should the value be?

Reply:
> Can we assume potentialValues contains only one value?

I think we can assume that. Since we do not allow the last element of the path to be glob, we only have two scenarios for location:
1. `spec.containers[name: *].image` - `recurse()` will hit the end of the path twice because there are two images. Each `recurse()` will return an string array of length 1, containing the respective image (my-docker-image and my-sidercar-image)
2. `spec.containers[name:my-docker-image]` - once we hit the end of the path, we select the container spec based on the key-value (`name:my-docker-image`). We can safely assume that each container must have a unique name so `recurse()` will return an string array of length 1.

> what should the value be?

`[]string{"my-docker-image", "my-sidercar-image"}`

Reply:
Just read your comments below. Based on it, let's only accept scenario 1 for now then.

Reply:
I agree the base case for each recursion should yield only one value. My question would be, after we've finished recursing and have gathered our values (per your function signature, this could be a list of strings), how do you re-associate those values with the correct fields to be replaced?

If I've convinced you that we can remove this function in my other comment, then no need to worry about the question, since it'd be mooted.

Reply:
After processing all placeholders, we should have a `map[ProviderName]map[OutboundData]Value`. Each placeholder object will use its provider name and value at location field to select the response from the map. Would that work?

Reply:
I think so!